### PR TITLE
Make user/group lookup caching thread-safe

### PR DESCRIPTION
**Pull request by pmatilai**
This seems like a huge overkill when in 4.19.1 there's exactly one rpmfiStat() call that unnecessarily invokes an rpmug lookup in a threaded scenario, but then rpmfiStat() and rpmfilesStat() are public APIs that people expect to be safe for use within the originating thread.

Collect all the caching data into a struct and allocate on per-thread basis, this seems like the path of least trouble in this case and git diff --stat agrees.

It's worth noting that this also simplifies the caching, we no longer keep separate name to id and id to name, totaling caches totaling four. We simply cache one id/name for for user and another for group data. Also, reset ids to 0 rather than -1, this is far more obviously a safe value as we have special cases to handle id 0.

rpmugUname() and rpmugGname() are have no users in the current codebase, so this was developed wrt commit c576d96930bfaf8f4328170de4c2915e76c1f715 where they still are used, to avoid nasty surprises later on if somebody finds a new case for these.

Fixes: #2826

### CODE REVIEW

**Comment by pmatilai** [comment]
(commit message tweaked somewhat in the later pushes)

**Review by Conan-Kudo — APPROVED** [review]

**Comment by ffesti** [comment]
I wonder if this leaks memory. We create a new struct for each thread. Yes, this is freed in rpmChrootSet but that is relying  a lot on the right call order.

**Comment by pmatilai** [comment]
It does leak, should've mentioned that in the commit message. In the sense that each thread calling it will leave one cache around and reachable. It's far from ideal but it's basically an emergency bandaid.

What do you mean about ending itself?

**Comment by ffesti** [comment]
OK, "ending itself" is a bit over dramatic... rpmChrootSet instantly returns if there is now chroot so rpmugFree() is never called at all.

**Comment by pmatilai** [comment]
Yup, but this patch doesn't change that, the "leak" is already there. This only makes it per-thread.

**Comment by pmatilai** [comment]
It wouldn't be hard to introduce per-thread locking instead. The bigger deal here is the new central struct that makes it possible to do stuff, perhaps I should actually split that into a separate commit and then consider the thread safety separately. The reason its lumped into one is basically wanting a single patch to fix the issue and people to cherry-pick, but that's clearly bending other "rules" here.

**Comment by pmatilai** [comment]
Oh, except that rpmugGname() and rpmugUname() additionally rely on central storage of the returned values so simple mutex locking doesn't work for those. So those would have to be changed to return malloced data for a simple fix (it wont break any users because there aren't any, at the moment).

**Comment by pmatilai** [comment]
Split to refactor + thread-safe fix, and restored that rpmugFree() on librpm exit that had gone missing at some point.

**Comment by ffesti** [comment]
Looks good to me.

File: lib/rpmug.c [hunk 1]
```diff
@@ -129,119 +143,97 @@ static int lookup_str(const char *path, long val, int vcol, int rcol,
...
     return 0;
 }
 
 const char * rpmugUname(uid_t uid)
 {
-    static uid_t lastUid = (uid_t) -1;
-    static char * lastUname = NULL;
-
-    if (uid == (uid_t) -1) {
-	lastUid = (uid_t) -1;
```
Comment (dmnks) [inline lib/rpmug.c:191]: I wonder if this line shouldn't be kept? `rpmugFree()` won't reset `lastUid` to `-1` and it can be used before assignment.

**Comment by dmnks** [comment]
It seems like the commit hash mentioned in the commit message isn't correct (it's the "Bump CI" commit which doesn't seem to have anything to do with this).

File: lib/rpmug.c (hunk 1 above)
Comment (dmnks) [reply lib/rpmug.c:191]: Ah, this is not needed now that we have a `struct`, of course. Nevermind the above :sweat_smile:

**Review by dmnks — APPROVED** [review]
Other than the commit note above, the patches look good to me.

**Review by Conan-Kudo — APPROVED** [review]
Modulo note as @dmnks stated, it looks good to me.

**Comment by pmatilai** [comment]
> It seems like the commit hash mentioned in the commit message isn't correct (it's the "Bump CI" commit which doesn't seem to have anything to do with this).

It's actually explained in the message:

>  rpmugUname() and rpmugGname() are have no users in the current codebase,
> so this was developed wrt commit c576d96930bfaf8f4328170de4c2915e76c1f715
> where they still are used

Ie, that commit is the one before the one that removed all rpmugUname() and rpmugGname() users, so I had to develop against that to be sure this rework actually works.

**Pull request merged**